---
title: "`sceptre` at scale"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{`sceptre` at scale}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

This vignette shows how to run `sceptre` at scale in a Linux or Unix environment.

# Important note

`sceptre` is powered under the hood by `ondisc`, an R package that enables fast, universal, and intuitive computing on large-scale single-cell data. If you would like to apply at-scale `sceptre` to your own data, you must (i) convert your gene expression and gRNA perturbation data into `ondisc_matrix` objects, (ii) produce a list of gene-gRNA pairs to analyze, and (iii) compute the cell-specific matrix of technical factors. Please see the tutorial "[Setting up a single-cell pooled CRISPR screen](https://timothy-barry.github.io/ondisc/articles/setting_up_a_crispr_screen.html)," located on the `ondisc` [website](https://github.com/timothy-barry/ondisc), for instructions on how to do this.

If you have completed the above tutorial, you can work through the entirety of this vignette. If not, you can work through sections 1-3 of this vignette, which demonstrate how to run `sceptre` at scale on a small, example dataset that ships with the package.

# 1. Obtaining the scripts

First, ensure that you have downloaded and installed `sceptre`.

```{r, eval=FALSE}
install.packages("devtools")
devtools::install_github("timothy-barry/sceptre")
```

Next, open the terminal and navigate to a directory in which you would like to run the analysis. For example:

```
# in terminal
mkdir ~/my_sceptre_dir
cd ~/my_sceptre_dir
```

Execute the following command to copy the scripts required to run `sceptre` into the current directory. Be sure to include the period at the end.

```
# in terminal
pkg_dir=$(Rscript -e "cat(find.package('sceptre'))"); cp -r $pkg_dir"/at_scale_scripts" .
```

This command creates a directory called `at_scale_scripts` containing a bash script `sceptre_at_scale.bash` and an R file `param_file.R`. The bash script `sceptre_at_scale.bash` contains code to run the analysis, and the R file `param_file.R` stores the parameters to be used in the analysis.

```
# in terminal
cd at_scale_scripts
ls # view sceptre_at_scale.bash and param_file.R
```

# 2. Examining the parameter file

Let's look at the contents of `param_file.R`:

```{r}
#param_file.R
library(ondisc)
library(sceptre)
param_funct <- function(param) {
        switch(param,
               # modify the parameters below
               storage_dir = "~/my_sceptre_dir/",
               expression_matrix = ondisc_matrix(system.file("extdata",
                                                             "expressions.h5",
                                                             package = "sceptre")),
               perturbation_matrix = ondisc_matrix(system.file("extdata",
                                                               "perturbations.h5",
                                                               package = "sceptre")),
               covariate_matrix = readRDS(system.file("extdata",
                                                      "covariate_matrix.rds",
                                                      package = "sceptre")),
               gene_gRNA_pairs = readRDS(system.file("extdata",
                                                     "gene_gRNA_pairs.rds",
                                                     package = "sceptre")),
               side = "left",
               pod_sizes = c(gene = 3, gRNA = 2, pair = 5),
               regularization_amount = 1,
               seed = 4,
               B = 500)
}
```

This script contains a single function `param_funct`, which takes the name of a parameter as an argument and returns the value of that parameter. `param_funct` accepts the following arguments:

- `storage_dir`: the name of a directory in which to store the intermediate computations and final results.

- `expression_matrix`: the gene expression data, represented as an `ondisc_matrix`.

- `perturbation_matrix`: the gRNA perturbation data, represented as an `ondisc_matrix`.

- `covariate_matrix`: the matrix of cell-specific covariates, represented as a data frame.

- `gene_gRNA_pairs`: a data frame storing the gene-gRNA pairs to analyze. The data frame must contain columns named `gene_id` and `gRNA_id`.

- `side`: the sidedness of the test; "left" performs a left-tailed test, "right" performs a right-tailed test, and "both" performs a two-tailed test. "left" is most appropriate for experiments in which cis-regulatory relationships are tested by perturbing putative enhancers with CRISPRi.

- `pod_sizes`: an integer vector giving the size of the "gene", "gRNA", and "pair" pods. `sceptre` groups the genes, gRNAs, and gene-gRNA pairs into distinct "pods" and runs computations on these pods in parallel. Smaller pod sizes give rise to greater parallelization. `pod_sizes` is vector of length three with names "gene", "gRNA", and "pair" and integer values giving the pod sizes.

- `regularization_amount`: the amount of regularization to apply to the estimated negative binomial size parameters, where 0 corresponds to no regularization at all. A reasonable default choice is 1. Note that `regularization_amount` affects only the power, not the validity, of the test.

- `B`: number of random samples to draw in the conditional randomization test; a reasonable default choice is 500.

- `seed`: seed to pass to the random number generator; any integer value suffices.

By default, `param_file.R` contains example data that ship with `sceptre`. `expression_matrix` and `perturbation_matrix` are small gene expression and gRNA perturbation matrices, respectively, sampled from Gasperini et al. 2019.

```{r}
expression_matrix <- ondisc_matrix(system.file("extdata",
                                 "expressions.h5",
                                 package = "sceptre"))
expression_matrix
perturbation_matrix <- ondisc_matrix(system.file("extdata",
                                   "perturbations.h5",
                                   package = "sceptre"))
perturbation_matrix
```

`covariate_matrix` stores the cell-specific technical factors for `expression_matrix` and `perturbation_matrix`.
 
```{r}
covariate_matrix <- readRDS(system.file("extdata",
                                "covariate_matrix.rds",
                                package = "sceptre"))
head(covariate_matrix)
```

Finally, `gene_gRNA_pairs` contains a list of gene-gRNA pairs for which we seek *p*-values.

```{r}
gene_gRNA_pairs <-  readRDS(system.file("extdata",
                               "gene_gRNA_pairs.rds",
                               package = "sceptre"))
head(gene_gRNA_pairs)
```

# 3. Running `sceptre` on the example data

To run `sceptre` on the set of parameters defined in `param_file.R`, simply call `bash` on the file `sceptre_at_scale.bash` within the directory `at_scale_scripts`.

```
# in terminal
bash sceptre_at_scale.bash
```

Alternately, make the file `sceptre_at_scale.bash` executable and execute it directly.

```
# in terminal
chmod 700 sceptre_at_scale.bash
./sceptre_at_scale.bash
```

Executing this script results in the creation of several new subdirectories in the `storage_dir` directory: `gene_precomp`, `gRNA_precomp`, `logs`, and `results`. `gene_precomp` and `gRNA_precomp` store the results of the intermediate computations; `logs` stores the log files that allow us to monitor the progress of the program; and `results` stores the results, including the final result file `all_results.rds`.

We can view the `logs` files during program execution by opening a new terminal and navigating to the appropriate directory.

```
# in new terminal window
cd ~/my_sceptre_dir/logs
# view gene_precomp_round_1_pod_1.Rout, for example
cat gene_precomp_round_1_pod_1.Rout
```

When the program has finished running, we can load the final results into R for downstream analysis.

```{r,eval=FALSE}
results <- readRDS(file = paste0(storage_dir, "/results/all_results.rds"))
```

# 4. Running `sceptre` on other datasets

We can apply `sceptre` to arbitrary single-cell CRISPR screen datasets. For example, we can apply `sceptre` to the Gasperini et al. 2019 data that we preprocessed in the tutorial "[Setting up a single-cell pooled CRISPR screen](https://timothy-barry.github.io/ondisc/articles/setting_up_a_crispr_screen.html)." Recall that we created several files in that tutorial, all of which we saved to the "processed data" directory:

1. The expressions.rds `ondisc_matrix`, plus its backing .h5 file,
2. The perturbations.rds `ondisc_matsrix`, plus its backing .h5 file,
3. The cell_covariate_matrix.rds file, and
4. The gene_gRNA_pairs.rds file.

To run `sceptre` on these data, we must create an new parameter file. We begin by copying the contents of `param_file.R` into a new file `param_file_big.R`.

```
# in terminal
cd ~/my_sceptre_dir/at_scale_scripts
cp param_file.R param_file_big.R
```

Next, we open the file `param_file_big.R` and update the fields withinin the `switch` statement. Be sure to change the file path of `example_dir` so that it matches the `example_dir` you defined in "Setting up a single-cell pooled CRISPR screen." (See the first code block in Section 1 of that tutorial.)

```{r,eval=FALSE}
# param_file_big.R
library(ondisc)
example_dir <- "/Volumes/tims_new_drive/research/sceptre_example" # change me!

processed_dir <- paste0(example_dir, "/processed_data")
library(ondisc)
param_funct <- function(param) {
        switch(param,
               # modify the parameters below
               storage_dir = paste0(example_dir, "/storage_dir"),
               expression_matrix = readRDS(paste0(processed_dir, "/expressions.rds")),
               perturbation_matrix = readRDS(paste0(processed_dir, "/perturbations.rds")),
               covariate_matrix = readRDS(paste0(processed_dir, "/cell_covariate_matrix.rds")),
               gene_gRNA_pairs = readRDS(paste0(processed_dir, "/gene_gRNA_pairs.rds")),
               side = "left",
               pod_sizes = c(gene = 100, gRNA = 5, pair = 100),
               regularization_amount = 1,
               seed = 4,
               B = 500)
}
```

Notice that we have changed the location of `storage_dir` to a subdirectory in `example_dir`. For large-scale analyses, it is often convenient put the "storage directory" nearby the data, which in this case are located in `example_dir`.


We run `sceptre` on the parameters defined in this file by calling `bash` on `sceptre_at_scale.bash`, passing `param_file_big.R` as a command line argument.

```
# in terminal
bash sceptre_at_scale.bash param_file_big.R
```

This program takes about 15 minutes to run on a Macbook Pro. There might be several warnings produced by "vglm.fitter" about convergence. These warnings are related to the gene expression regressions and are safe to ignore, as the validity of `sceptre` does not depend on an accurate model for gene expression. There additionally might be several warnings produced by "st.infoUv" about "score not quite 0." These warnings are related to the skew-t fit of the resampled test statistics and likewise are safe to ignore, as the score (i.e., gradient of the log-likelihood) does not need to be exactly 0 for the *p*-value to be accurate.

The results of the analysis are stored in example_dir/storage_dir/results.

```{r, eval=FALSE}
results_fp <- paste0(example_dir, "/storage_dir/results/all_results.rds")
results <- readRDS(results_fp)
```

And with that, we have completed our analysis!
