---
title: "Running sceptre on moderately sized data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Running sceptre on moderately sized data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup,message=FALSE}
library(dplyr)
library(sceptre)
```

This tutorial shows how to run `sceptre` on a dataset of moderate size. The data should fit into memory and be small enough such that the analysis can run on a single machine (as opposed to a computer cluster). To keep the size of the package small, we generate synthetic data rather than use real data.

## Obtain (or generate) the raw data

First, we generate a gene-by-cell expression matrix and a binary gRNA-by-cell perturbation matrix. We use 40 genes, 50 gRNAs, and 5000 cells. Genes and gRNAs are put in rows, and cells are put in columns, as is standard in single-cell analysis.

```{r}
set.seed(4)
# n genes, gRNAs, and cells
n_genes <- 40
n_gRNAs <- 50
n_cells <- 5000
# gene expression matrix 
expression_matrix <- replicate(n = n_genes, rpois(n_cells, 1)) %>% t()
# perturbation matrix
perturbation_matrix <- replicate(n = n_gRNAs, rbinom(n_cells, 1, 0.05)) %>% t()
```

The "expression" and "perturbation" matrices must have row names corresponding to the genes and gRNAs, respectively. We assign simple names to each.

```{r}
row.names(expression_matrix) <- paste0("gene", seq(1, nrow(expression_matrix)))
row.names(perturbation_matrix) <- paste0("gRNA", seq(1, nrow(perturbation_matrix)))
```

Finally, for each cell, we simulate its sequencing batch: "batch 1" or "batch 2."

```{r}
batch <- sample(x = c("batch1", "batch2"), size = n_cells, replace = TRUE, prob = c(0.5, 0.5))
```

On a real dataset we would perform quality control at this stage. For example, we would remove cells with 0 perturbations. We recommend doing minimal quality control, as `sceptre` is robust and can handle unusual gene expression distributions and misspecified expression models.

## Compute the cell-wise covariate matrix

Next, we compute the cell-specific covariate matrix. The following covariates are reasonable to include:

1. sequencing batch
2. percentage of UMIs that map to mitochondrial genes
3. log-transformed mRNA UMI count
4. log-transformed number of genes expressed in cell
5. log-transformed number of perturbations detected in cell, or log-transfored gRNA UMI count

In this simple tutorial we will use (1) and (3). First, we compute the log-transformed mRNA UMI counts.

```{r}
lg_mRNA_lib_size <- log(colSums(expression_matrix))
```

Next, we create the cell-specific covariate matrix, putting cells in rows and covariates in columns.

```{r}
covariate_matrix <- data.frame(lg_mRNA_lib_size = lg_mRNA_lib_size,
                               batch = factor(batch))
```

## Select the genes and gRNAs to include in the analysis

After this, we select the gene-gRNA pairs to analyze. If we were conducting a *cis* analysis, we might pair each gene with all gRNAs targeting within 1Mb of the gene's transcription start site. For simplicity, we will randomly sample 90 gene-gRNA pairs. The data frame that we create, `gene_gRNA_pairs`, should have one column called `gene_id` and a second column called `gRNA_id`.

```{r}
gene_gRNA_pairs <- expand.grid(gene_id = row.names(expression_matrix),
                               gRNA_id = row.names(perturbation_matrix)) %>% slice_sample(n = 90)
```

## Run `sceptre`

The final step is to run `sceptre`. We use the function `run_sceptre_in_memory`, which, as the name suggests, runs `sceptre` on in-memory expression and perturbation matrices. Notice that we did not "normalize" the gene expression data at any step; `sceptre` operates on raw UMI count matrices.

`run_sceptre_in_memory` takes several arguments, which we summarize below.

- `storage_dir`: the name of a directory in which to store the intermediate computations and final results.

- `expression_matrix`: the gene-by-cell expression data, represented as an R matrix. The names of the rows should be the gene names.

- `perturbation_matrix`: the gRNA-by-cell perturbation data, represented as an R matrix. The names of the rows should be the gRNA names.

- `covariate_matrix`: the matrix of cell-specific covariates, represented as a data frame. The rows correspond to cells, and the columns to the covariates.

- `gene_gRNA_pairs`: a data frame storing the gene-gRNA pairs to analyze. The data frame must contain columns named `gene_id` and `gRNA_id`, with entries corresponding to the row names of `expression_matrix` and `perturbation_matrix`.

- `side`: the sidedness of the test: one of "left," "right," and "both." "left" is appropriate when perturbations are expected to decrease the expression of genes.

- `pod_sizes`: an integer vector giving the size of the "gene", "gRNA", and "pair" pods. `run_sceptre_in_memory` groups genes, gRNAs, and gene-gRNA pairs into distinct "pods" and runs computations on these pods in parallel. Smaller pod sizes give rise to greater parallelization. `pod_sizes` should be an integer vector of length three, with values giving the pod sizes and names "gene," "gRNA," and "pair."

- `regularization_amount`: the amount of regularization to apply to the estimated negative binomial size parameters, where 0 corresponds to no regularization at all. The default option is 0.1.

- `B`: the number of random samples to draw in the conditional randomization test. The default option is 500.

- `seed`: seed to pass to the random number generator. The default value is 4.

We define `storage_dir`, `side`, and `pod_sizes`, which we have not yet set. We will use the default options for the remaining parameters.

```{r}
storage_dir <- tempdir()
side <- "left"
pod_sizes <- c(gene = 10, gRNA = 10, pair = 15)
```

We have set `storage_dir` to `tempdir` for the sake of portability; in general, you should choose a more meaningful directory to store the intermediate computations and final results.

Finally, we call the function.

```{r,cache=TRUE}
result <- run_sceptre_in_memory(storage_dir,
                     expression_matrix,
                     perturbation_matrix,
                     covariate_matrix,
                     gene_gRNA_pairs,
                     side,
                     pod_sizes)
```

As the function runs, we can check the `logs` subdirectory of the storage directory to monitor progress.

```
# in terminal
cd /var/folders/h6/pt59hmwd7rqb3jqfzwh0hlth0000gn/T//RtmpUfO717 # update path
cd logs
```

This function takes about 5 minutes to execute on a Macbook Pro laptop with four Intel Core i5 processors.

## Interpreting the results

Let's examine the results (which also have been written to the `results` subdirectory of the storage directory).

```{r}
head(result)
```

The data frame contains the following columns:

- `gene_id`: the ID of the gene

- `gRNA_id`: the ID of the gRNA

- `p_value`: the gRNA-to-gene *p*-value

- `skew_t_fit_success`: boolean indicating whether the skew-t fit was successful (if not, the empirical conditional randomization test *p*-value is returned)

- `xi`, `omega`, `alpha`, `nu`: parameters of the fitted skew-t distribution

- `z_value`: the z-score of the negative binomial regression on the observed (i.e., non-resampled) data. The sign `z_value` gives the direction of the effect.

- `n_successful_resamples`: the number of resamples in the conditional randomization test that were succesful

We might want to apply a Benjamini-Hochberg correction to account for multiple hypothesis testing, but be sure to proceed with caution here.

```{r}
result$p_value_adj <- p.adjust(p = result$p_value, method = "BH")
```

At this point we have finished our analysis!
