---
title: "Applying `sceptre` to small a small dataset"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Applying `sceptre` to small a small dataset}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


This tutorial demonstrates how to apply `sceptre` to a small dataset consisting of a single gene and gRNA. We use example data taken from the study [A Genome-wide Framework for Mapping Gene Regulation via Cellular Genetic Screens](https://www.cell.com/cell/fulltext/S0092-8674(18)31554-X?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS009286741831554X%3Fshowall%3Dtrue), published by Gasperini et al. in 2019. The gene (ID: ENSG00000164713) and gRNA (ID: chr7.3255_top_two) that we investigate were selected randomly from the set of all genes and gRNAs in the study.

When analyzing a large number of genes and gRNAs, it is better to use the approach outlined in the tutorial  [`sceptre` at scale](https://timothy-barry.github.io/sceptre/articles/sceptre-at-scale.html). The large-scale approach has the advantage of (i) regularizing the negative binomial regression test statistic by pooling information across genes, and (ii) recycling certain repetitive computations, thereby accelerating program execution. The small-scale approach outlined here is appropriate only for datasets consisting of a handful of genes and gRNAs.

# Setup and data

First, ensure that you have downloaded and installed `sceptre`.

```{r, eval=FALSE}
install.packages("devtools")
devtools::install_github("timothy-barry/sceptre")
```

Load the package and example data.

```{r, message=FALSE}
library(sceptre)
data("expressions")
data("gRNA_indicators")
data("covariate_matrix")
```

The *expressions* vector contains the gene expressions (in number of UMIs); the *gRNA_indicator* vector records whether the gRNA was detected or not detected in a given cell; and the *covariate_matrix* records cell-specific technical factors (in this case, (i) fraction of UMIs that map to mitochondrial genes, (ii) sequencing batch, (iii) log-transformed total UMI count across genes, (iv) log-transformed total number of detected gRNAs, and (v) log-transformed total genes expressed). There are 205,797 cells in the data. Let's take a peek at the data by calling *head*.

```{r}
head(expressions)
head(gRNA_indicators)
head(covariate_matrix)
```

# Applying SCEPTRE

Next, we run SCEPTRE by calling the function `run_sceptre_gRNA_gene_pair`. This function returns a *p*-value corresponding to the null hypothesis that the perturbation does not impact gene expression, accounting for the technical factors.

```{r run-algo,cache=TRUE}
result <- run_sceptre_gRNA_gene_pair(expressions = expressions, 
                                      gRNA_indicators = gRNA_indicators, 
                                      covariate_matrix = covariate_matrix,
                                      B = 1000,
                                      seed = 1234,
                                      reduced_output = FALSE)
```

We can examine the resultant *p*-value.

```{r}
result$p_value
```

We also can examine the negative binomial z-score and the parameters of the fitted skew-t distribution used to compute the *p*-value.

```{r}
result$z_value
result$skew_t_mle
```

Finally, we can create a histogram of the resampled, null z-scores.

```{r, fig.retina = 2, fig.align="center", fig.width=5, fig.height=4}
p <- plot_skew_t(resampled_zvalues = result$resampled_z_values, original_zvalue = result$z_value, dp = result$skew_t_mle, interval = c(-4,4))
plot(p)
```

Given that the computed *p*-value (*p* = 0.21) exceeds our pre-chosen significance threshold of 0.05, we retain our null hypothesis that the perturbation does not affect gene expression. In practice, when analyzing a large dataset, we would apply a Benjamini-Hochberg correction to all gene-gRNA pairs and use a more conservative threshold to determine significance.
