---
title: "Basic `sceptre` tutorial"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{using_sceptre}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette illustrates application of `sceptre` to a small, example, high multiplicity of infection (MOI) single-cell CRISPR screen dataset. We begin by installing and loading all required packages, including the up-to-date version of `sceptre`.

```{r, eval=FALSE}
install.packages("devtools")
devtools::install_github("katsevich-lab/sceptre")
install.packages("tidyverse")
install.packages("cowplot")
```

```{r, message=FALSE, warning=FALSE}
library(tidyverse)
library(cowplot)
library(Matrix)
library(sceptre)
```

## Step 1: Prepare the data

The first step is to prepare the data to pass to `sceptre`. We must prepare three separate data objects: the gene-by-cell expression matrix, the gRNA-by-cell expression matrix, and the cell-specific matrix of covariates.

### Gene and gRNA expression matrices

We load the example gene-by-cell and gRNA-by-cell expression matrices that ship with the `sceptre` package.

```{r, source_data}
data(gene_matrix)
data(gRNA_matrix)
```

One can use the commands `?gene_matrix` and `?gRNA_matrix` to learn more about these data. Briefly, `gene_matrix` (respectively, `gRNA_matrix`) is a 5 x 106,660 (respectively, 50 x 106,660) matrix of gene (respectively, gRNA) unique molecular identifier (UMI) counts. The data are taken from the paper "Global analysis of enhancer targets reveals convergent enhancer-driven regulatory modules" by Xie et al., 2019. The authors used a CRISPRi-based assay to target putative enhancers in a population of K562 cells. Both genes and gRNAs are downsampled here to reduce the size of the data.

The row names of `gene_matrix` and `gRNA_matrix` are the gene IDs and gRNA IDs, respectively. The column names, meanwhile, are the cell barcodes.

```{r}
# row names
row.names(gene_matrix)
row.names(gRNA_matrix) %>% head()

# column names
colnames(gene_matrix) %>% head()
colnames(gRNA_matrix) %>% head()
```

To get a better feel for the data, we plot a histogram of the expressions of an arbitrarily selected gene ("ENSG00000105819.13") and gRNA ("AGAAGAGTCAGCCTTGCAC").

```{r, plot_hist, fig.width=7, fig.height=3, fig.retina=2}
example_gene <- gene_matrix["ENSG00000105819.13",]
example_gRNA <- gRNA_matrix["AGAAGAGTCAGCCTTGCAC",]

hist_gene <- ggplot(data = example_gene %>% tibble(count = .) %>% filter(count >= 1, count <= 15), mapping = aes(x = count)) +
  geom_histogram(binwidth = 2, col = "black", fill = "dodgerblue3", alpha = 0.7) +
  scale_y_continuous(trans='log10', expand = c(0, NA)) + xlab("Gene expressions") + ylab("") + theme_bw(base_size = 10)

hist_gRNA <- ggplot(data = example_gRNA %>% tibble(count = .) %>% filter(count >= 1, count <= 40), mapping = aes(x = count)) +
  geom_histogram(binwidth = 2, col = "black", fill = "orchid4", alpha = 0.7) +
  scale_y_continuous(trans='log10', expand = c(0, NA)) + xlab("gRNA expressions") + ylab("") + theme_bw(base_size = 10)

plot_grid(hist_gene, hist_gRNA, labels = c("a", "b"))
```

As expected, the data are highly discrete counts. Note that we do *not* normalize either the gene or gRNA expression matrices before passing them to `sceptre`.

### Cell-wise covariate matrix

Next, we load the cell-wise covariate matrix.

```{r}
data(covariate_matrix)
head(covariate_matrix)
```

`covariate_matrix` is a 106,660 x 4 data frame of "technical factors" or covariates. The row names are the cell barcodes, and the column names are the covariates. The covariates are as follows:

-   Log-transformed gene library size (`lg_gene_lib_size`); can be computed via `log(colSums(gene_matrix))`

-   Log-transformed gRNA library size (`lg_gRNA_lib_size`); can be computed via `log(colSums(gRNA_matrix))`

-   Sequencing batch (`batch`)

-   Percentage of gene transcripts that map to mitochondrial genes (`p_mito`)

**We strongly recommend that users include these same covariates in their cell-wise covariate matrix**.

## Step 2: (Optional) Combine gRNAs that target the same site

The second step, which is optional, is to combine gRNAs that target the same chromosomal site. We load `gRNA_grps`, a list containing information about the chromosomal position that each gRNA targets.

```{r}
data(site_table)
```

```{r}
gRNA_matrix_collapsed <- combine_gRNAs(gRNA_matrix, site_table)
row.names(gRNA_matrix_collapsed)
```

## Step 3: Determine which gRNA-gene pairs to analyze

```{r}
data(gene_gRNA_pairs)
head(gene_gRNA_pairs)
```

## Step 4: Determine the sidedness of the test

| Target element  | CRISPR modality  |     Testing for...     | Sidedness |
|:---------------:|:----------------:|:----------------------:|:---------:|
| Enhancer or TSS | Inhibition or KO | Decrease in expression |   Left    |
|    Silencer     | Inhibition or KO | Increase in expression |   Right   |
| Enhancer or TSS |    Activation    | Increase in expression |   Right   |
|    Silencer     |    Activation    | Decrease in expression |   Left    |

## Step 5: Run the method

```{r, cache=TRUE, eval=FALSE}
result <- run_sceptre_high_moi(gene_matrix = gene_matrix,
                               gRNA_matrix = gRNA_matrix_collapsed,
                               covariate_matrix = covariate_matrix,
                               gene_gRNA_pairs = gene_gRNA_pairs,
                               side = "left")
```

```{r,eval=FALSE}
head(result)
```

`?plot_result` for a function to plot the result.
