<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Applying `sceptre` to small a small dataset • sceptre</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Applying `sceptre` to small a small dataset">
<meta property="og:description" content="sceptre">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sceptre</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Function reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/sceptre-at-scale.html">sceptre at scale</a>
    </li>
    <li>
      <a href="../articles/sceptre-small-example.html">sceptre on small data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="sceptre-small-example_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Applying <code>sceptre</code> to small a small dataset</h1>
            
      
      
      <div class="hidden name"><code>sceptre-small-example.Rmd</code></div>

    </div>

    
    
<p>This tutorial demonstrates how to apply <code>sceptre</code> to a small dataset consisting of a single gene and gRNA. We use example data taken from the study <a href="https://www.cell.com/cell/fulltext/S0092-8674(18)31554-X?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS009286741831554X%3Fshowall%3Dtrue">A Genome-wide Framework for Mapping Gene Regulation via Cellular Genetic Screens</a>, published by Gasperini et al. in 2019. The gene (ID: ENSG00000164713) and gRNA (ID: chr7.3255_top_two) that we investigate were selected randomly from the set of all genes and gRNAs in the study.</p>
<p>When analyzing a large number of genes and gRNAs, it is better to use the approach outlined in the tutorial <a href="https://timothy-barry.github.io/sceptre/articles/sceptre-at-scale.html"><code>sceptre</code> at scale</a>. The large-scale approach has the advantage of (i) regularizing the negative binomial regression test statistic by pooling information across genes, and (ii) recycling certain repetitive computations, thereby accelerating program execution. The small-scale approach outlined here is appropriate only for datasets consisting of a handful of genes and gRNAs.</p>
<div id="setup-and-data" class="section level1">
<h1 class="hasAnchor">
<a href="#setup-and-data" class="anchor"></a>Setup and data</h1>
<p>First, ensure that you have downloaded and installed <code>sceptre</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"timothy-barry/sceptre"</span><span class="op">)</span></code></pre></div>
<p>Load the package and example data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">sceptre</span><span class="op">)</span>
<span class="fu"><a href="../reference/data.html">data</a></span><span class="op">(</span><span class="st">"expressions"</span><span class="op">)</span>
<span class="fu"><a href="../reference/data.html">data</a></span><span class="op">(</span><span class="st">"gRNA_indicators"</span><span class="op">)</span>
<span class="fu"><a href="../reference/data.html">data</a></span><span class="op">(</span><span class="st">"covariate_matrix"</span><span class="op">)</span></code></pre></div>
<p>The <em>expressions</em> vector contains the gene expressions (in number of UMIs); the <em>gRNA_indicator</em> vector records whether the gRNA was detected or not detected in a given cell; and the <em>covariate_matrix</em> records cell-specific technical factors (in this case, (i) fraction of UMIs that map to mitochondrial genes, (ii) sequencing batch, (iii) log-transformed total UMI count across genes, (iv) log-transformed total number of detected gRNAs, and (v) log-transformed total genes expressed). There are 205,797 cells in the data. Let’s take a peek at the data by calling <em>head</em>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">expressions</span><span class="op">)</span>
<span class="co">#&gt; [1] 0 3 2 4 0 2</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">gRNA_indicators</span><span class="op">)</span>
<span class="co">#&gt; [1] 0 0 0 0 0 0</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">covariate_matrix</span><span class="op">)</span>
<span class="co">#&gt;        p_mito   prep_batch lg_total_umis lg_guide_count lg_n_genes</span>
<span class="co">#&gt; 1 0.058786706 prep_batch_1      9.774062       4.204693   8.174421</span>
<span class="co">#&gt; 2 0.036086518 prep_batch_1      9.096387       3.258097   7.841100</span>
<span class="co">#&gt; 3 0.069823051 prep_batch_1      9.591308       4.110874   8.068090</span>
<span class="co">#&gt; 4 0.026186508 prep_batch_1     10.034428       3.663562   8.420462</span>
<span class="co">#&gt; 5 0.007991318 prep_batch_1      9.223849       3.610918   7.865188</span>
<span class="co">#&gt; 6 0.022356681 prep_batch_1      9.185125       4.043051   7.690286</span></code></pre></div>
</div>
<div id="applying-sceptre" class="section level1">
<h1 class="hasAnchor">
<a href="#applying-sceptre" class="anchor"></a>Applying SCEPTRE</h1>
<p>Next, we run SCEPTRE by calling the function <code>run_sceptre_gRNA_gene_pair</code>. This function returns a <em>p</em>-value corresponding to the null hypothesis that the perturbation does not impact gene expression, accounting for the technical factors.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_sceptre_gRNA_gene_pair.html">run_sceptre_gRNA_gene_pair</a></span><span class="op">(</span>expressions <span class="op">=</span> <span class="va">expressions</span>, 
                                      gRNA_indicators <span class="op">=</span> <span class="va">gRNA_indicators</span>, 
                                      covariate_matrix <span class="op">=</span> <span class="va">covariate_matrix</span>,
                                      B <span class="op">=</span> <span class="fl">1000</span>,
                                      seed <span class="op">=</span> <span class="fl">1234</span>,
                                      reduced_output <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Running gRNA precomputation.</span>
<span class="co">#&gt; Running gene precomputation.</span>
<span class="co">#&gt; Running resample 100/1000.</span>
<span class="co">#&gt; Running resample 200/1000.</span>
<span class="co">#&gt; Running resample 300/1000.</span>
<span class="co">#&gt; Running resample 400/1000.</span>
<span class="co">#&gt; Running resample 500/1000.</span>
<span class="co">#&gt; Running resample 600/1000.</span>
<span class="co">#&gt; Running resample 700/1000.</span>
<span class="co">#&gt; Running resample 800/1000.</span>
<span class="co">#&gt; Running resample 900/1000.</span>
<span class="co">#&gt; Running resample 1000/1000.</span></code></pre></div>
<p>We can examine the resultant <em>p</em>-value.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">result</span><span class="op">$</span><span class="va">p_value</span>
<span class="co">#&gt; [1] 0.2138241</span></code></pre></div>
<p>We also can examine the negative binomial z-score and the parameters of the fitted skew-t distribution used to compute the <em>p</em>-value.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">result</span><span class="op">$</span><span class="va">z_value</span>
<span class="co">#&gt; [1] -0.3037842</span>
<span class="va">result</span><span class="op">$</span><span class="va">skew_t_mle</span>
<span class="co">#&gt;            xi         omega         alpha            nu </span>
<span class="co">#&gt; -6.563831e-02  1.242149e+00  8.000418e-01  1.185023e+06</span></code></pre></div>
<p>Finally, we can create a histogram of the resampled, null z-scores.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_skew_t.html">plot_skew_t</a></span><span class="op">(</span>resampled_zvalues <span class="op">=</span> <span class="va">result</span><span class="op">$</span><span class="va">resampled_z_values</span>, original_zvalue <span class="op">=</span> <span class="va">result</span><span class="op">$</span><span class="va">z_value</span>, dp <span class="op">=</span> <span class="va">result</span><span class="op">$</span><span class="va">skew_t_mle</span>, interval <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></code></pre></div>
<p><img src="sceptre-small-example_files/figure-html/unnamed-chunk-7-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Given that the computed <em>p</em>-value (<em>p</em> = 0.21) exceeds our pre-chosen significance threshold of 0.05, we retain our null hypothesis that the perturbation does not affect gene expression. In practice, when analyzing a large dataset, we would apply a Benjamini-Hochberg correction to all gene-gRNA pairs and use a more conservative threshold to determine significance.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://timothy-barry.github.io/">Timothy Barry</a>, Eugene Katsevich.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
